File: 3507. Minimum Pair Removal to Sort Array I.cpp
#include <vector>
#include <climits>
#include <algorithm>

class Solution {
public:
    bool check(const std::vector<int>& arr) {
        for (size_t i = 1; i < arr.size(); i++)
            if (arr[i] < arr[i - 1])
                return false;
        return true;
    }

    int minimumPairRemoval(const std::vector<int>& nums) {
        std::vector<int> arr(nums.begin(), nums.end());

        while (!check(arr)) {
            int sum = INT_MAX;
            for (size_t i = 1; i < arr.size(); i++)
                sum = std::min(sum, arr[i] + arr[i - 1]);

            std::vector<int> changedArr;
            bool merged = false;

            for (size_t i = 0; i < arr.size();) {
                if (!merged && i + 1 < arr.size() && arr[i] + arr[i + 1] == sum) {
                    changedArr.push_back(arr[i] + arr[i + 1]);
                    merged = true;
                    i += 2;
                } else {
                    changedArr.push_back(arr[i]);
                    i++;
                }
            }

            arr = std::move(changedArr);
        }

        return static_cast<int>(nums.size() - arr.size());
    }
};

File: 3507. Minimum Pair Removal to Sort Array I.java
class Solution {
    public boolean check(ArrayList<Integer> arr) {
        for (int i = 1; i < arr.size(); i++)
            if (arr.get(i) < arr.get(i - 1))
                return false;

        return true;
    }

    public int minimumPairRemoval(int[] nums) {
        ArrayList<Integer> arr = new ArrayList<>();
        for (int num : nums)
            arr.add(num);

        while (!check(arr)) {
            int sum = Integer.MAX_VALUE;
            for (int i = 1; i < arr.size(); i++)
                sum = Math.min(sum, arr.get(i) + arr.get(i - 1));

            ArrayList<Integer> changedArr = new ArrayList<>();
            boolean merged = false;

            for (int i = 0; i < arr.size();) {
                if (!merged && i + 1 < arr.size()
                        && arr.get(i) + arr.get(i + 1) == sum) {
                    changedArr.add(arr.get(i) + arr.get(i + 1));
                    merged = true;
                    i += 2;
                } else {
                    changedArr.add(arr.get(i));
                    i++;
                }
            }

            arr = changedArr;
        }

        return nums.length - arr.size();
    }
}